tiledb/sm/compressors/delta_compressor.cc

```diff
@@ -0,0 +1,307 @@
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements the delta compressor class.
+ */
+
+#include "tiledb/sm/compressors/delta_compressor.h"
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/buffer/buffer.h"
+#include "tiledb/sm/enums/datatype.h"
+
+/* ****************************** */
+/*             MACROS             */
+/* ****************************** */
+
+#define ABS(x) ((x) < 0) ? uint64_t(-(x)) : uint64_t(x)
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+const uint64_t Delta::OVERHEAD = 17;
+
+/* ****************************** */
+/*               API              */
+/* ****************************** */
+
+void Delta::compress(
+    Datatype type, ConstBuffer* input_buffer, Buffer* output_buffer) {
+  switch (type) {
+    case Datatype::BLOB:
+      Delta::compress<std::byte>(input_buffer, output_buffer);
+      break;
+    case Datatype::INT8:
+      Delta::compress<int8_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::BOOL:
+    case Datatype::UINT8:
+      Delta::compress<uint8_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::INT16:
+      Delta::compress<int16_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::UINT16:
+      Delta::compress<uint16_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::INT32:
+      Delta::compress<int>(input_buffer, output_buffer);
+      break;
+    case Datatype::UINT32:
+      Delta::compress<uint32_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::INT64:
+      Delta::compress<int64_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::UINT64:
+      Delta::compress<uint64_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::CHAR:
+      Delta::compress<char>(input_buffer, output_buffer);
+      break;
+    case Datatype::DATETIME_YEAR:
+    case Datatype::DATETIME_MONTH:
+    case Datatype::DATETIME_WEEK:
+    case Datatype::DATETIME_DAY:
+    case Datatype::DATETIME_HR:
+    case Datatype::DATETIME_MIN:
+    case Datatype::DATETIME_SEC:
+    case Datatype::DATETIME_MS:
+    case Datatype::DATETIME_US:
+    case Datatype::DATETIME_NS:
+    case Datatype::DATETIME_PS:
+    case Datatype::DATETIME_FS:
+    case Datatype::DATETIME_AS:
+    case Datatype::TIME_HR:
+    case Datatype::TIME_MIN:
+    case Datatype::TIME_SEC:
+    case Datatype::TIME_MS:
+    case Datatype::TIME_US:
+    case Datatype::TIME_NS:
+    case Datatype::TIME_PS:
+    case Datatype::TIME_FS:
+    case Datatype::TIME_AS:
+      Delta::compress<int64_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::STRING_ASCII:
+    case Datatype::STRING_UTF8:
+    case Datatype::STRING_UTF16:
+    case Datatype::STRING_UTF32:
+    case Datatype::STRING_UCS2:
+    case Datatype::STRING_UCS4:
+    case Datatype::ANY:
+      Delta::compress<uint8_t>(input_buffer, output_buffer);
+      break;
+    case Datatype::FLOAT32:
+    case Datatype::FLOAT64:
+      throw StatusException(
```

Comment:
Did we want to support the casting here?

Reply:
We're going to see the reinterpret datatype here, if it's set: https://github.com/TileDB-Inc/TileDB/pull/3969/files#diff-f46f30278b5b0f9992c18dab02d507dc5f489bc304d82a6404c692da761ddaf3R365